internal/exec/stages/files/filesystemEntries.go

```diff
@@ -236,10 +237,21 @@ func (s stage) getOrderedCreationList(config types.Config) ([]filesystemEntry, e
 		}
 		paths[path] = l.Path
 		l.Path = path
-		entries = append(entries, linkEntry(l))
+		isHardLink := l.Hard != nil && *l.Hard
+		if !isHardLink {
+			entries = append(entries, linkEntry(l))
+		} else {
+			hardlinks = append(hardlinks, linkEntry(l))
+		}
+
 	}
 	sort.Slice(entries, func(i, j int) bool { return util.Depth(entries[i].node().Path) < util.Depth(entries[j].node().Path) })
 
+	// Append all the hard links to the list after the sort function.
+	// For more information: https://github.com/coreos/ignition/issues/800
+	for _, hardlink := range hardlinks {
+		entries = append(entries, hardlink)
```

Comment:
Rather than looping you can append it like so `entries = append(entries, hardlinks...)`